valhalla/baldr/datetime.h

```diff
@@ -19,7 +19,10 @@
 #include <valhalla/baldr/nodeinfo.h>
 #include <valhalla/midgard/constants.h>
 
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
 #include <valhalla/proto/tripcommon.pb.h>
+#pragma GCC diagnostic pop
```

Comment:
this is so unmanageable though. we'll have to litter this all over the code. maybe we could make a macro out of it or something:

`INCLUDE_PROTO valhalla/proto/tripcommon.pb.h`

OR maybe we should make a separate cmake target with all the generated protobuffer stuff in it and include just one header everywhere?

OR we could just disable this warning globally?

none of the options are particularly great

Reply:
> this is so unmanageable though
agreed

I don't really like macro for a few reasons. I was thinking about upgrading cmake step. As I see all proto-headers live in `${BUILD_DIR}/src/valhalla/` directory. Maybe we could add
```
for header in ${BUILD_DIR}/src/valhalla/*.h:
  insert "#pragma GCC diagnostic push\n#pragma GCC diagnostic ignored "-Wunused-parameter" to the beginning of the header
  insert `#pragma GCC diagnostic pop` to the end of the header
```

I'll take a few days to think about best approach

Reply:
`SYSTEM` option in https://cmake.org/cmake/help/latest/command/target_include_directories.html looks like the light at the end of the tunnel
`
If SYSTEM is specified, the compiler will be told the directories are meant as system include directories on some platforms (signalling this setting might achieve effects such as the compiler skipping warnings...
`

Reply:
I looked at this too for a similar PR but couldn't quite get it to work when I tried. I'll follow this PR to see if you find a way.

Reply:
I did a few unsuccessful tries to disable warnings in external headers using `SYSTEM`  cmake option. Results:
The best I got is disabling warnings in all headers(external + internal/valhalla headers): I inserted `SYSTEM` right before `MODULE_INCLUDE_DIRECTORIES` https://github.com/valhalla/valhalla/blob/master/src/CMakeLists.txt#L100. However, we are in cpp world and some part of our codebase is templated(i.e lives in headers). So don't think this is an option for us
In a toy example the `SYSTEM` option works. So we could create a wrapper for each external module and disable warnings for its headers, e.g
```cmake
add_library(protobuf_wrapper INTERFACE)
target_link_libraries(protobuf_wrapper protobuf)
target_include_directories(protobuf SYSTEM INTERFACE $<TARGET_PROPERTY:protobuf,INTERFACE_INCLUDE_DIRECTORIES>)
....

target_link_libraries(valhalla-mjolnir 
  protobuf_wrapper
  valhalla-midgard
...)
// each library defines in its own cmakelists whether it's headers are system or not
```

However, in order to get this we need to refactor valhalla cmake configuration, i.e operate with cmake targets rather than use target properties like https://github.com/valhalla/valhalla/blob/master/src/CMakeLists.txt#L133. From my experiments when we extract `INTERFACE_INCLUDE_DIRECTORIES` property we loose `SYSTEM` property(if it was set)
```cmake 
add_library(lib_a INTERFACE)
# system is set
target_include_directories(lib_a SYSTEM INTERFACE ./headers/) 

add_library(lib_b INTERFACE)
target_link_libraries(lib_b INTERFACE lib_a)
// doesn't work for some reason. system is lost?
target_include_directories(lib_b INTERFACE $<TARGET_PROPERTY:lib_a,INTERFACE_INCLUDE_DIRECTORIES>)
```

Reply:
> this is so unmanageable though. we'll have to litter this all over the code. maybe we could make a macro out of it or something:
> 
> `INCLUDE_PROTO valhalla/proto/tripcommon.pb.h`
> 
> OR maybe we should make a separate cmake target with all the generated protobuffer stuff in it and include just one header everywhere?
> 
> OR we could just disable this warning globally?
> 
> none of the options are particularly great

Finally, my choice is disabling unused-parameter for all sources. As it is the only warning that external headers have - its disabling solves the issue. I hope it's a short/mid term solution until cmake is refactored and we can apply `SYSTEM` cmake option